src/Raven.Client/Documents/Session/SessionInfo.cs

```diff
@@ -10,53 +10,62 @@ public class SessionInfo
         [ThreadStatic]
         private static int _clientSessionIdCounter;
 
-        private int _clientSessionId;
+        private int? _sessionId;
         private bool _sessionIdUsed;
         private readonly int _loadBalancerContextSeed;
+        private readonly bool _loadBalancerContextSelectorSet;
+        private readonly InMemoryDocumentSessionOperations _session;
 
         public int SessionId
         {
             get
             {
+                if (_sessionId == null)
+                    SetContext(_session.Conventions.LoadBalancerPerSessionContextSelector?.Invoke(_session.DatabaseName));
+
                 _sessionIdUsed = true;
-                return _clientSessionId;
+
+                return _sessionId.Value;
             }
         }
 
+        internal bool HasSessionId
+        {
+            get => _sessionId.HasValue || _loadBalancerContextSelectorSet;
+        }
+
         public long? LastClusterTransactionIndex { get; set; }
 
         public bool AsyncCommandRunning { get; set; }
 
         public bool NoCaching { get; set; }
 
-        public SessionInfo(string sessionKey, int loadBalancerContextSeed, bool asyncCommandRunning, long? lastClusterTransactionIndex = null, bool noCaching = false)
+        internal SessionInfo(InMemoryDocumentSessionOperations session, SessionOptions options, DocumentStoreBase documentStore, bool asyncCommandRunning)
         {
-            SetContext(sessionKey, loadBalancerContextSeed);
+            if (documentStore is null)
+                throw new ArgumentNullException(nameof(documentStore));
 
-            _loadBalancerContextSeed = loadBalancerContextSeed;
+            _session = session ?? throw new ArgumentNullException(nameof(session));
+            _loadBalancerContextSeed = session.RequestExecutor.Conventions.LoadBalancerContextSeed;
+            _loadBalancerContextSelectorSet = session.Conventions.LoadBalanceBehavior == LoadBalanceBehavior.UseSessionContext && session.Conventions.LoadBalancerPerSessionContextSelector != null;
```

Comment:
I think this is wrong. If the admin explicitly turned it off, even if you provide `SetContext`, it should not apply it.